src/Microsoft.ApplicationInsights.AspNet/TelemetryInitializers/OperationNameTelemetryInitializer.cs

```diff
@@ -9,82 +9,98 @@
     using Microsoft.AspNet.Http;
     using Microsoft.AspNet.Mvc;
     using Microsoft.AspNet.Mvc.Routing;
+    using Microsoft.AspNet.Routing;
     using Microsoft.Framework.DependencyInjection;
+    using Microsoft.Framework.Notification;
 
     public class OperationNameTelemetryInitializer : TelemetryInitializerBase
     {
-        public OperationNameTelemetryInitializer(IHttpContextAccessor httpContextAccessor) : base(httpContextAccessor)
-        { }
+        public const string BeforeActionNotificationName = "Microsoft.AspNet.Mvc.BeforeAction";
+
+        public OperationNameTelemetryInitializer(IHttpContextAccessor httpContextAccessor, INotifier notifier) 
+            : base(httpContextAccessor)
+        {
+            if (notifier == null)
+            {
+                throw new ArgumentNullException("notifier");
+            }
+
+            notifier.EnlistTarget(this);
+        }
 
         protected override void OnInitializeTelemetry(HttpContext platformContext, RequestTelemetry requestTelemetry, ITelemetry telemetry)
         {
             if (string.IsNullOrEmpty(telemetry.Context.Operation.Name))
             {
-                string name = this.GetNameFromRouteContext(platformContext.RequestServices);                
-
-                if (string.IsNullOrEmpty(name))
+                if (!string.IsNullOrEmpty(requestTelemetry.Name))
                 {
-                    name = platformContext.Request.Path.Value;
+                    telemetry.Context.Operation.Name = requestTelemetry.Name;
                 }
-
-                name = platformContext.Request.Method + " " + name; 
-                
-                var telemetryType = telemetry as RequestTelemetry;
-                if (telemetryType != null && string.IsNullOrEmpty(telemetryType.Name))
+                else
                 {
-                    telemetryType.Name = name;
+                    // We didn't get BeforeAction notification
+                    string name = platformContext.Request.Method + " " + platformContext.Request.Path.Value;
+                    requestTelemetry.Name = name;
+                    requestTelemetry.Context.Operation.Name = name;
+                    telemetry.Context.Operation.Name = name;
                 }
+            }
+        }
+
+        [NotificationName(BeforeActionNotificationName)]
+        public void OnBeforeAction(ActionDescriptor actionDescriptor, HttpContext httpContext, RouteData routeData)
+        {
+            string name = this.GetNameFromRouteContext(routeData);
+            var telemetry = httpContext.RequestServices.GetService<RequestTelemetry>();
 
-                telemetry.Context.Operation.Name = name;
+            if (!string.IsNullOrEmpty(name) && telemetry != null)
+            {
+                name = httpContext.Request.Method + " " + name;
+                ((RequestTelemetry)telemetry).Name = name;
+                ((RequestTelemetry)telemetry).Context.Operation.Name = name;
```

Comment:
no need to set it - it will be set by the same telemetry initializer when it will be called for RequestTelemetry